api/client/client.go

```diff
@@ -0,0 +1,254 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/textileio/go-textile-core/crypto/symmetric"
+	pb "github.com/textileio/go-textile-threads/api/pb"
+	es "github.com/textileio/go-textile-threads/eventstore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Client provides the client api
+type Client struct {
+	client pb.APIClient
+	ctx    context.Context
+	cancel context.CancelFunc
+	conn   *grpc.ClientConn
+}
+
+// ListenEvent is used to send data or error values for Listen
+type ListenEvent struct {
+	data []byte
+	err  error
+}
+
+// NewClient starts the client
+func NewClient(host string, port int) (*Client, error) {
+	url := fmt.Sprintf("%v:%v", host, port)
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	client := &Client{
+		client: pb.NewAPIClient(conn),
+		ctx:    ctx,
+		cancel: cancel,
+		conn:   conn,
+	}
+	return client, nil
+}
+
+// Close closes the client's grpc connection and cancels any active requests
+func (c *Client) Close() error {
+	c.cancel()
+	return c.conn.Close()
+}
+
+// NewStore cereates a new Store
+func (c *Client) NewStore() (string, error) {
+	resp, err := c.client.NewStore(c.ctx, &pb.NewStoreRequest{})
+	if err != nil {
+		return "", err
+	}
+	return resp.GetID(), nil
+}
+
+// RegisterSchema registers a new model shecma
+func (c *Client) RegisterSchema(storeID, name, schema string) error {
+	req := &pb.RegisterSchemaRequest{
+		StoreID: storeID,
+		Name:    name,
+		Schema:  schema,
+	}
+	_, err := c.client.RegisterSchema(c.ctx, req)
+	return err
+}
+
+// Start starts the specified Store
+func (c *Client) Start(storeID string) error {
+	_, err := c.client.Start(c.ctx, &pb.StartRequest{StoreID: storeID})
+	return err
+}
+
+// StartFromAddress starts the specified Store with the provided address and keys
+func (c *Client) StartFromAddress(storeID string, addr ma.Multiaddr, followKey, readKey *symmetric.Key) error {
+	req := &pb.StartFromAddressRequest{
+		StoreID:   storeID,
+		Address:   addr.String(),
+		FollowKey: followKey.Bytes(),
+		ReadKey:   readKey.Bytes(),
+	}
+	_, err := c.client.StartFromAddress(c.ctx, req)
+	return err
+}
+
+// ModelCreate creates new instances of model objects
+func (c *Client) ModelCreate(storeID, modelName string, items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.ModelCreateRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		Values:    values,
+	}
+
+	resp, err := c.client.ModelCreate(c.ctx, req)
+	if err != nil {
+		return err
+	}
+
+	for i, entity := range resp.GetEntities() {
+		err := json.Unmarshal([]byte(entity), items[i])
```

Comment:
Related to the conversation below about whether the store should return bytes or strings, if here you just returned an error you wouldn't give the user the impression that they can expect the client to unmarshal model instances. Alternatively, the find methods could return `[]interface{}` and just have to perform a type assertion. Right now it's sorta both: "i'll unmarshal when i can, otherwise you do it"

Reply:
I'm not sure if you're saying things are fine as is or they need to be changed? Like, for example, you say "if here you just returned an error...". Here I do just return and error, but it's not clear if you are saying that is good or bad. Let's chat through this more.

Reply:
Oop, sorry, that was confusing. I was considering the unmarshalled items as part of the return... forget that, I'll rephrase: If you let the user do the unmarshalling, they wouldn't have the expectation that the other methods would also perform unmarshalling, and it would feel more appropirate to deal with `[][]byte` in the client api (not referring to the internal JSON mode API). Though, looking again, this unmarshalling here is just to include the `ID` param, making it useful to keep.